Feature/react router with redux

guidsen opened the pull request:
Made react router with redux work. Add a base route with a base reducer with some dummy items.
Only problem left is the fact that the event handlers aren't being fired.
Probably missed something..

ericelliott commented on source/client/index.js:
```diff
@@ -1,13 +1,18 @@
-const props = window.payload || {
-  title: 'Default client title'
+import { Provider } from 'react-redux';
+import store from '../shared/configureStore';
```
should be:

```js
import store from 'shared/configureStore';
```

ericelliott commented on source/server/render-layout.js:
```diff
@@ -1,13 +1,12 @@
     <html>
       <head>
-        <title>${ title }</title>
+        <title>React App Hodor</title>
```
Shouldn't we pass in the document title? It definitely shouldn't be "React App Hodor".

ericelliott commented on source/server/render-layout.js:
```diff
@@ -1,13 +1,12 @@
       </head>
       <body>
         <div id='root'>${ rootMarkup }</div>
-        <script>${ payload }</script>
```
How do we bootstrap client data?

guidsen commented:
Thanks for the feedback @ericelliott. I've updated the PR. The event handlers still won't fire though. Try to click the button on the screen, the `onClick` doesn't fire in `shared/components/App.js`.

ericelliott commented:
Have you seen [redux-simple-router](https://github.com/jlongster/redux-simple-router)? See the [blog post for details](http://jlongster.com/A-Simple-Way-to-Route-with-Redux).

guidsen commented:
@ericelliott Yeah I saw it. cool, so easy to implement as well. Would you mind checking out what could be wrong with the event handlers not firing like I mentioned earlier?

ericelliott commented on package.json:
```diff
@@ -35,11 +35,13 @@
     "babel-eslint": "4.1.3",
     "babel-loader": "5.4.0",
-    "babel-plugin-react-transform": "1.1.1",
+    "babel-plugin-react-transform": "^1.1.1",
```
Please don't unpin these. See ["Software Versions are Broken"](https://medium.com/javascript-scene/software-versions-are-broken-3d2dc0da0783#.ce8a3737u) for details.

ericelliott replied:
Any plans to remove these carets?

guidsen replied:
Sorry, forgot about them. Thanks for the link to the resource.

ericelliott commented on source/server/routes/main/index.js:
```diff
@@ -1,23 +1,26 @@
-import settings from 'server/settings';
+import { match } from 'react-router';
+import routes from '../../../shared/routes';
```
Should be:

```js
import routes from 'shared/routes';
```

Please don't use relative paths.

guidsen replied:
Does NPM always look from the root directory?

ericelliott replied:
See [Requiring Modules](https://github.com/cloverfield-tools/universal-react-boilerplate#requiring-modules).

guidsen replied:
Cool!

ericelliott commented on source/server/settings/index.js:
```diff
@@ -1,5 +1,5 @@
 
 export default Object.assign({}, appHome, {
-  TITLE: 'Your App'
+  TITLE: 'Your Cool React App!'
```
I like the enthusiasm! 🎸🎶

ericelliott commented on source/shared/components/App.js:
```diff
@@ -0,0 +1,30 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+
+class App extends Component {
```
NONONONO. No.

guidsen replied:
Haha you like pure functions more, right...? :P

ericelliott replied:
For components that aren't pure functions, use [react-stamp](https://github.com/stampit-org/react-stamp).

ericelliott commented on source/shared/components/App.js:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  render () {
+    return <button onClick={this.addReader.bind(this)}>Test</button>;
```
I take it this is what you're having trouble with? Generally speaking, you don't need `.bind()` in ES6. Try this?

```js
return <button onClick={ () => this.addReader(book) }>Test</button>;
```

But going further, you probably don't even need the `.addReader()` method:

```js
render ({dispatch}) {
  return <button onClick={ () => dispatch({ type: 'ADD_COUNT', book }); }>Test</button>;
}
```

ericelliott commented on webpack.config.dev.js:
```diff
@@ -17,13 +17,29 @@ module.exports = {
   plugins: [
     new webpack.HotModuleReplacementPlugin(),
-    new webpack.NoErrorsPlugin()
+    new webpack.NoErrorsPlugin(),
```
Did this file really need to change at all?

ericelliott commented on source/client/index.js:
```diff
@@ -1,13 +1,22 @@
+
+const render = reactDom.renderToString;
+
+const store = configureStore(window.__INITIAL_STATE__);
```
These underscores are causing lint failures. Maybe just give it a more semantic name, instead? Maybe...

```js
const store = configureStore(window.BOOTSTRAP_CLIENT_STATE);
```

Remember that the server code needs to be updated as well.

ericelliott commented:
Please fix lint failures. =)

ericelliott commented:
This build is failing because you deleted the title component, and there is still a test trying to use it. The correct solution to this problem is to put back the title component (which should not have been deleted. It's an example of how to organize components).

You've also renamed `shared/components/app/index.js` to App.js -- did you have a good reason to do that?

ericelliott commented:
Please note that you could avoid checking in broken builds if you run `npm run watch` in your terminal and fix any build failures prior to committing your changes. =)

guidsen commented:
@ericelliott I renamed `shared/components/app/index.js` to `App.js`, because I think it's easier to understand a common used pattern for beginners. I'm not sure where it would make sense to have a sub-folder with a index.js file for a component.

And the reason why I've removed the `React` parameter for the Title component is because I think it will be easier to understand for new users as well.

nkbt commented on source/shared/components/App.js:
```diff
@@ -0,0 +1,31 @@
+    return (
+      <div key={ book.id }>
+        { book.text } - Read by { book.count } people.
+        <button onClick={ () => dispatch({ type: 'ADD_COUNT', book }) }>Add reader</button>
```
Generally using anonymous function as prop is not good since every time render() is called this function is created again and again.
Better "cache" it as instance method or in case of pure component - as function outside of component (if possible) or another pure component.

See https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md for examples and explanation.

ericelliott replied:
I wouldn't worry about this unless it is actually identified as a performance issue in the application. Maybe if you're making a video game and animating thousands of components on the screen at 60fps...

nkbt commented on webpack.config.dev.js:
```diff
@@ -22,8 +22,24 @@ module.exports = {
-      include: path.join(__dirname, 'source')
+      include: path.join(__dirname, 'source'),
+      loader: 'babel',
+      query: {
```
Better to use .babelrc for this.

If you want particular config to be applied only for dev env, then use `env` in`.babelrc`:

```json
{
  "env": {
    "development": {
       "plugins": []
     }
  }
}
```

https://babeljs.io/docs/usage/babelrc

guidsen replied:
I've tried that, but seems when added the `query` data inside the `.babelrc` file, I got `react-transform` errors. I've seen some other people having this issue as well and it was fixed by adding it directly to the webpack loader query.

nkbt replied:
I meant not to use `query` for babel loader at all. Combining loader `query` and `.babelrc` is definitely not something I would recommend.

guidsen replied:
I've deleted the `.babelrc`. .babelrc doesn't work with react-transform like I tested.

nkbt replied:
The problem with query approach - you can't have multiple loaders.

It could be solved in one or the other way, but is very confusing for people.

`.babelrc` is for babel only, it should not affect any other tools. And it should work the same way as with query params. What was the problem with `react-transform`?

ericelliott commented:
> I renamed shared/components/app/index.js to App.js, because I think it's easier to understand a common used pattern for beginners.

Beginner materials often trivialize things to the point that it's not useful. The priority in this repo is to demonstrate good practices, such as consistently organizing and consistently naming components. =)

guidsen commented:
@ericelliott I've updated it to a factory. It looks and feels better to have a function indeed. What are the benefits of having a factory function to instantiate a component like we do now?

ericelliott commented:
> What are the benefits of having a factory function to instantiate a component like we do now?

See [this article](https://medium.com/@dan_abramov/two-weird-tricks-that-fix-react-7cf9bbdef375). Having multiple React instances in the same app can cause problems, and users may want to load React from a CDN for the client side. Exporting factories allows them to load React from CDN an inject a single React instance for the app to share.

nkbt commented on package.json:
```diff
@@ -40,6 +40,8 @@
     "eslint-plugin-react": "3.6.3",
     "express": "4.13.3",
+    "react-transform-catch-errors": "1.0.0",
+    "react-transform-hmr": "1.0.1",
```
By the way it does not support babel6. As well as https://github.com/gaearon/babel-plugin-react-transform.
Not an issue here, but worth knowing.

ericelliott replied:
I can upgrade this to Babel 6 later.

guidsen commented on source/shared/routes.js:
```diff
@@ -0,0 +1,8 @@
+export default React => () => {
+  const App = createApp(React);
+
+  return <Route path="/" component={App} />;
```
@ericelliott This doesn't work for some reason. Am I missing something?

ericelliott replied:
Define "doesn't work".

guidsen replied:
The routes aren't returned the expected object for the server side rendering.
When I remove the factory like this:

```
import React from 'react';
import { Route } from 'react-router';
import createApp from './components/App';

const App = createApp(React);

export default (
  <Route path="/" component={App} />
);
```

I get a `Cannot call class as a function` error.

Changed `import createRoutes from 'shared/routes';` to `import routes from 'shared/routes';` as well.

Everything runs fine without errors, but the `App` component isn't being loaded. It shows a `Not found` message (see `server/routes/main/index.js`)

ericelliott replied:
I don't have time to take a deep look at this right now, but there's a [full example of redux-simple-router usage](https://github.com/freeqaz/redux-simple-router-example) you can use for reference. Have you looked at that?

ericelliott commented:
I'd like to use [redux-simple-router](https://github.com/jlongster/redux-simple-router) for this, and we really should apply this update **after** the [Babel 6 upgrade](https://github.com/cloverfield-tools/universal-react-boilerplate/issues/57).

Want to reopen after those changes?

vasco3 commented:
I would say that for the next PR to keep it tight to the scope of the issue.. so it's easier/quicker to merge.

ericelliott commented:
The idea of this PR is on track. We just need to get it working, and make sure we apply it on top of all the latest modules to make it easier. =)

guidsen commented:
Ill try to make the routing work later today. Ill reopen after the Babel 6 upgrade ;).
